log4j-api/src/main/java/org/apache/logging/log4j/spi/AbstractLogger.java

```diff
@@ -2734,4 +2748,108 @@ public void warn(final String message, final Object p0, final Object p1, final O
             final Object p7, final Object p8, final Object p9) {
         logIfEnabled(FQCN, Level.WARN, null, message, p0, p1, p2, p3, p4, p5, p6, p7, p8, p9);
     }
+
+    /**
+     * Constuct a trace log event.
+     * @return a LogBuilder.
+     * @since 3.0
+     */
+    @Override
+    public  LogBuilder atTrace() {
+        if (isTraceEnabled()) {
+            return logBuilder.get().setLevel(Level.TRACE);
```

Comment:
I think we need something similar to `ReusableParameterizedMessage.reserved` for this to work, otherwise we will clobber threadlocal state with something like this:

```java
private void func() {
  log.atInfo().withMessage("Value {}").withParameters(getValue()).log();
}

private Object getValue() {
    log.atInfo().withMessage("Getting Value").log();
    return value;
}
```

Reply:
Ouch. I hadn't thought of that. That means the threadLocal will need to maintain a list of LogBuilders instead of just one.

Reply:
Could we keep a single builder with a `reserved` flag? If the current value is reserved, we can remove it from the threadlocal and create a new instance for the nested logger. I think that works well for ReusableParameterizedMessage, and limits the per-thread memory we can retain. It think nested log statements are uncommon enough that we don't need to make them allocation free, though they may be more common using a builder API than the standard style APIs.